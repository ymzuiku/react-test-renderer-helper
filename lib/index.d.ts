import { ReactElement, ReactType } from 'react';

interface ReactTestRendererJSON {
  type: string;
  props: { [propName: string]: any };
  children: null | ReactTestRendererJSON[];
}
interface ReactTestRendererTree extends ReactTestRendererJSON {
  nodeType: 'component' | 'host';
  instance: any;
  rendered: null | ReactTestRendererTree;
}
interface ReactTestInstance {
  instance: any;
  type: ReactType;
  props: { [propName: string]: any };
  parent: null | ReactTestInstance;
  children: Array<ReactTestInstance | string>;

  find(predicate: (node: ReactTestInstance) => boolean): ReactTestInstance;
  findByType(type: ReactType): ReactTestInstance;
  findByProps(props: { [propName: string]: any }): ReactTestInstance;

  findAll(predicate: (node: ReactTestInstance) => boolean, options?: { deep: boolean }): ReactTestInstance[];
  findAllByType(type: ReactType, options?: { deep: boolean }): ReactTestInstance[];
  findAllByProps(props: { [propName: string]: any }, options?: { deep: boolean }): ReactTestInstance[];
}

interface ReactTestRenderer {
  toJSON(): null | ReactTestRendererJSON;
  toTree(): null | ReactTestRendererTree;
  unmount(nextElement?: ReactElement<any>): void;
  update(nextElement: ReactElement<any>): void;
  getInstance(): null | ReactTestInstance;
  root: ReactTestInstance;
}
interface ISnapshot {
  status: any;
  state: any;
  props: { [propName: string]: any };
  zip: Object;
}

export function find(component: ReactTestRenderer, id: string): ReactTestInstance;
export function getStatus(key: any): any;
export function clearStatus(key: any): void;
export function toSnapshotByHooks(component: ReactTestRenderer): ISnapshot;
export function toSnapshot(component: ReactTestRenderer, ...any): ISnapshot;
