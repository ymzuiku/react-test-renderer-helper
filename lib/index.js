!function(e,t){if("object"===typeof exports&&"object"===typeof module)module.exports=t();else if("function"===typeof define&&define.amd)define([],t);else{var r=t();for(var n in r)("object"===typeof exports?exports:e)[n]=r[n]}}(window,function(){return function(e){var t={};function r(n){if(t[n])return t[n].exports;var o=t[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,r),o.l=!0,o.exports}return r.m=e,r.c=t,r.d=function(e,t,n){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},r.r=function(e){"undefined"!==typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&"object"===typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(r.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var o in e)r.d(n,o,function(t){return e[t]}.bind(null,o));return n},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,"a",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p="/ymzuiku/react-test-renderer-helper.git/",r(r.s=2)}([function(e,t,r){"use strict";t.__esModule=!0,window.__testerCache={},window.__testerIsTest=!0;var n="tester__defaultKey";var o={setStatus:function(e,t){var r,o;window.__testerIsTest&&("undefined"===typeof t?(o=n,r=e):(o=e,r=t),window.__testerCache[o]=r)},getStatus:function(e){return"undefined"===typeof e&&(e=n),window.__testerCache[e]},clearStatus:function(e){"undefined"===typeof e&&(e=n),"*"===e?window.__testerCache={}:delete window.__testerCache[e]},status:window.__testerCache};t.default=o,e.exports=t.default},function(e,t){e.exports=require("react-test-renderer")},function(e,t,r){"use strict";function n(e,t,r){return t in e?Object.defineProperty(e,t,{value:r,enumerable:!0,configurable:!0,writable:!0}):e[t]=r,e}r.r(t);var o=r(1),u=r.n(o),i=r(0),f=r.n(i);r.d(t,"fixId",function(){return s}),r.d(t,"find",function(){return c}),r.d(t,"getStatus",function(){return p}),r.d(t,"clearStatus",function(){return l}),r.d(t,"toSnapshotByHooks",function(){return d}),r.d(t,"toSnapShot",function(){return y}),window.__testerIsTest=!0;var a=u.a.create().toJSON();function s(e){var t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:"";if(e&&e.props){if(t.search(/\=/)>0){var r=t.split("=");return e.props[r[0]]===r[1]}if(0===t.search(/\#/)){var n=t.split("#");return e.props.id===n[1]}if(0===t.search(/\@/)){var o=t.split("@");return e.props.name===o[1]}if(0===t.search(/\./)){var u=t.split(".");return e.props.className===u[1]}return e.type===t}return!1}function c(e,t){var r,n,o=arguments.length>2&&void 0!==arguments[2]?arguments[2]:0,u=-1;return s(n=e&&e.toJSON?e.toJSON():e,t)&&(u+=1),u===o?a||n:(function e(n){if(n&&n.children)for(var i=0;i<n.children.length&&!r;i++){var f=n.children[i];s(f,t)&&(u+=1),u===o?r=f:e(f)}}(n),a||r)}function p(e){return f.a.getStatus(e)}function l(e){f.a.clearStatus(e)}function d(e,t){return!1===t?{props:e.root.props,zip:e.toJSON()}:{status:p(t),props:e.root.props,zip:e.toJSON()}}function y(e){for(var t=e.getInstance(),r={status:p(e),props:t.props,state:t.state,zip:e.toJSON()},o=arguments.length,u=new Array(o>1?o-1:0),i=1;i<o;i++)u[i-1]=arguments[i];if(u.length>0){for(var f={},a=0;a<u.length;a++)f[u[a]]=t[u[a]];return function(e){for(var t=1;t<arguments.length;t++){var r=null!=arguments[t]?arguments[t]:{},o=Object.keys(r);"function"===typeof Object.getOwnPropertySymbols&&(o=o.concat(Object.getOwnPropertySymbols(r).filter(function(e){return Object.getOwnPropertyDescriptor(r,e).enumerable}))),o.forEach(function(t){n(e,t,r[t])})}return e}({},r,{other:f})}return r}a=null}])});
//# sourceMappingURL=index.js.map