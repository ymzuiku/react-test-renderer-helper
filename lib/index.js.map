{"version":3,"sources":["../webpack/universalModuleDefinition","../webpack/bootstrap","lib/tester.lib.js","../external \"react-test-renderer\"","../node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/defineProperty.js","lib/index.lib.js","../node_modules/babel-preset-react-app/node_modules/@babel/runtime/helpers/esm/objectSpread.js"],"names":["root","factory","exports","module","define","amd","a","i","window","installedModules","__webpack_require__","moduleId","l","modules","call","m","c","d","name","getter","o","Object","defineProperty","enumerable","get","r","Symbol","toStringTag","value","t","mode","__esModule","ns","create","key","bind","n","object","property","prototype","hasOwnProperty","p","s","__testerCache","__testerIsTest","tester__defaultKey","tester__default","setStatus","b","obj","getStatus","clearStatus","status","default","require","_defineProperty","configurable","writable","rendererType","renderer","toJSON","fixId","ele","id","arguments","length","undefined","props","search","kv","split","className","type","find","component","child","rootComp","num","jumpNum","_getChild","comp","children","tester","toSnapshotByHooks","zip","toSnapshot","params","instance","getInstance","returnEle","state","other","target","source","ownKeys","keys","getOwnPropertySymbols","concat","filter","sym","getOwnPropertyDescriptor","forEach","_objectSpread"],"mappings":"CAAA,SAAAA,EAAAC,GACA,qBAAAC,SAAA,kBAAAC,OACAA,OAAAD,QAAAD,SACA,uBAAAG,eAAAC,IACAD,OAAA,GAAAH,OACA,CACA,IAAAK,EAAAL,IACA,QAAAM,KAAAD,GAAA,kBAAAJ,gBAAAF,GAAAO,GAAAD,EAAAC,IAPA,CASCC,OAAA,WACD,mBCTA,IAAAC,EAAA,GAGA,SAAAC,EAAAC,GAGA,GAAAF,EAAAE,GACA,OAAAF,EAAAE,GAAAT,QAGA,IAAAC,EAAAM,EAAAE,GAAA,CACAJ,EAAAI,EACAC,GAAA,EACAV,QAAA,IAUA,OANAW,EAAAF,GAAAG,KAAAX,EAAAD,QAAAC,IAAAD,QAAAQ,GAGAP,EAAAS,GAAA,EAGAT,EAAAD,QA0DA,OArDAQ,EAAAK,EAAAF,EAGAH,EAAAM,EAAAP,EAGAC,EAAAO,EAAA,SAAAf,EAAAgB,EAAAC,GACAT,EAAAU,EAAAlB,EAAAgB,IACAG,OAAAC,eAAApB,EAAAgB,EAAA,CAA0CK,YAAA,EAAAC,IAAAL,KAK1CT,EAAAe,EAAA,SAAAvB,GACA,qBAAAwB,eAAAC,aACAN,OAAAC,eAAApB,EAAAwB,OAAAC,YAAA,CAAwDC,MAAA,WAExDP,OAAAC,eAAApB,EAAA,cAAiD0B,OAAA,KAQjDlB,EAAAmB,EAAA,SAAAD,EAAAE,GAEA,GADA,EAAAA,IAAAF,EAAAlB,EAAAkB,IACA,EAAAE,EAAA,OAAAF,EACA,KAAAE,GAAA,kBAAAF,QAAAG,WAAA,OAAAH,EACA,IAAAI,EAAAX,OAAAY,OAAA,MAGA,GAFAvB,EAAAe,EAAAO,GACAX,OAAAC,eAAAU,EAAA,WAAyCT,YAAA,EAAAK,UACzC,EAAAE,GAAA,iBAAAF,EAAA,QAAAM,KAAAN,EAAAlB,EAAAO,EAAAe,EAAAE,EAAA,SAAAA,GAAgH,OAAAN,EAAAM,IAAqBC,KAAA,KAAAD,IACrI,OAAAF,GAIAtB,EAAA0B,EAAA,SAAAjC,GACA,IAAAgB,EAAAhB,KAAA4B,WACA,WAA2B,OAAA5B,EAAA,SAC3B,WAAiC,OAAAA,GAEjC,OADAO,EAAAO,EAAAE,EAAA,IAAAA,GACAA,GAIAT,EAAAU,EAAA,SAAAiB,EAAAC,GAAsD,OAAAjB,OAAAkB,UAAAC,eAAA1B,KAAAuB,EAAAC,IAGtD5B,EAAA+B,EAAA,2CAIA/B,IAAAgC,EAAA,kCChFAxC,EAAQ6B,YAAa,EAErBvB,OAAOmC,cAAgB,GACvBnC,OAAOoC,gBAAiB,EAExB,IAAIC,EAAqB,qBAiCzB,IAAIC,EAAkB,CACpBC,UAjCF,SAA2BzC,EAAG0C,GAI5B,IAAIC,EAAKf,EAHa,qBAAX1B,QAA2BA,OAAOoC,iBAI5B,qBAANI,GACTd,EAAMW,EACNI,EAAM3C,IAEN4B,EAAM5B,EACN2C,EAAMD,GAERxC,OAAOmC,cAAcT,GAAOe,IAsB5BC,UApBF,SAA2BhB,GAIzB,MAHmB,qBAARA,IACTA,EAAMW,GAEDrC,OAAOmC,cAAcT,IAiB5BiB,YAfF,SAA6BjB,GACR,qBAARA,IACTA,EAAMW,GAEI,MAARX,GACF1B,OAAOmC,cAAgB,GACvBG,EAAgBM,OAAS5C,OAAOmC,uBAEzBnC,OAAOmC,cAAcT,GAC5BY,EAAgBM,OAAS5C,OAAOmC,gBAOlCS,OAAQ5C,OAAOmC,eAGjBzC,EAAQmD,QAAUP,EAClB3C,EAAOD,QAAUA,EAAO,uBChDxBC,EAAAD,QAAAoD,QAAA,qDCAe,SAAAC,EAAAN,EAAAf,EAAAN,GAYf,OAXAM,KAAAe,EACA5B,OAAAC,eAAA2B,EAAAf,EAAA,CACAN,QACAL,YAAA,EACAiC,cAAA,EACAC,UAAA,IAGAR,EAAAf,GAAAN,EAGAqB,+RCVAzC,OAAOoC,gBAAiB,EAExB,IAAIc,EAAeC,IAAS1B,SAAS2B,SAG9B,SAASC,EAAMC,GAAc,IAATC,EAASC,UAAAC,OAAA,QAAAC,IAAAF,UAAA,GAAAA,UAAA,GAAJ,GAC9B,GAAIF,GAAOA,EAAIK,MAAO,CACpB,GAAIJ,EAAGK,OAAO,MAAQ,EAAG,CACvB,IAAMC,EAAKN,EAAGO,MAAM,KACpB,OAAOR,EAAIK,MAAME,EAAG,MAAQA,EAAG,GAEjC,GAAwB,IAApBN,EAAGK,OAAO,MAAa,CACzB,IAAMC,EAAKN,EAAGO,MAAM,KACpB,OAAOR,EAAIK,MAAMJ,KAAOM,EAAG,GAE7B,GAAwB,IAApBN,EAAGK,OAAO,MAAa,CACzB,IAAMC,EAAKN,EAAGO,MAAM,KACpB,OAAOR,EAAIK,MAAMjD,OAASmD,EAAG,GAE/B,GAAwB,IAApBN,EAAGK,OAAO,MAAa,CACzB,IAAMC,EAAKN,EAAGO,MAAM,KACpB,OAAOR,EAAIK,MAAMI,YAAcF,EAAG,GAEpC,OAAOP,EAAIU,OAAST,EAEtB,OAAO,EAGF,SAASU,EAAKC,EAAWX,GAAa,IACvCY,EAqBAC,EAtB8BC,EAASb,UAAAC,OAAA,QAAAC,IAAAF,UAAA,GAAAA,UAAA,GAAH,EAEpCc,GAAW,EA8Bf,OAHIjB,EALFe,EADEF,GAAaA,EAAUd,OACdc,EAAUd,SAEVc,EAGOX,KAClBe,GAAW,GAETA,IAAYD,EACPnB,GAAgBkB,GA9BzB,SAASG,EAAUC,GACjB,GAAIA,GAAQA,EAAKC,SACf,IAAK,IAAI1E,EAAI,EAAGA,EAAIyE,EAAKC,SAAShB,SAC5BU,EADoCpE,IAAK,CAI7C,IAAMuD,EAAMkB,EAAKC,SAAS1E,GACtBsD,EAAMC,EAAKC,KACbe,GAAW,GAETA,IAAYD,EACdF,EAAQb,EAERiB,EAAUjB,IAoBlBiB,CAAUH,GACHlB,GAAgBiB,GAGlB,SAASzB,EAAUhB,GACxB,OAAOgD,IAAOhC,UAAUhB,GAGnB,SAASiB,EAAYjB,GAC1BgD,IAAO/B,YAAYjB,GAGd,SAASiD,EAAkBT,EAAWX,GAC3C,OAAW,IAAPA,EACK,CACLI,MAAOO,EAAU1E,KAAKmE,MACtBiB,IAAKV,EAAUd,UAGZ,CACLR,OAAQF,EAAUa,GAClBI,MAAOO,EAAU1E,KAAKmE,MACtBiB,IAAKV,EAAUd,UAIZ,SAASyB,EAAWX,EAAWY,GACpC,IAAMC,EAAWb,EAAUc,cACrBC,EAAY,CAChBrC,OAAQF,EAAUwB,GAClBP,MAAOoB,EAASpB,MAChBuB,MAAOH,EAASG,MAChBN,IAAKV,EAAUd,UAEjB,GAAI0B,GAAUA,EAAOrB,OAAS,EAAG,CAE/B,IADA,IAAI0B,EAAQ,GACHpF,EAAI,EAAGA,EAAI+E,EAAOrB,OAAQ1D,IACjCoF,EAAML,EAAO/E,IAAMgF,EAASD,EAAO/E,IAErC,OCxGW,SAAAqF,GACf,QAAArF,EAAA,EAAiBA,EAAAyD,UAAAC,OAAsB1D,IAAA,CACvC,IAAAsF,EAAA,MAAA7B,UAAAzD,GAAAyD,UAAAzD,GAAA,GACAuF,EAAAzE,OAAA0E,KAAAF,GAEA,oBAAAxE,OAAA2E,wBACAF,IAAAG,OAAA5E,OAAA2E,sBAAAH,GAAAK,OAAA,SAAAC,GACA,OAAA9E,OAAA+E,yBAAAP,EAAAM,GAAA5E,eAIAuE,EAAAO,QAAA,SAAAnE,GACMqB,EAAcqC,EAAA1D,EAAA2D,EAAA3D,MAIpB,OAAA0D,EDwFWU,CAAA,GACFb,EADL,CAEEE,UAGJ,OAAOF,EAzGT/B,EAAe","file":"index.js","sourcesContent":["(function webpackUniversalModuleDefinition(root, factory) {\n\tif(typeof exports === 'object' && typeof module === 'object')\n\t\tmodule.exports = factory();\n\telse if(typeof define === 'function' && define.amd)\n\t\tdefine([], factory);\n\telse {\n\t\tvar a = factory();\n\t\tfor(var i in a) (typeof exports === 'object' ? exports : root)[i] = a[i];\n\t}\n})(window, function() {\nreturn "," \t// The module cache\n \tvar installedModules = {};\n\n \t// The require function\n \tfunction __webpack_require__(moduleId) {\n\n \t\t// Check if module is in cache\n \t\tif(installedModules[moduleId]) {\n \t\t\treturn installedModules[moduleId].exports;\n \t\t}\n \t\t// Create a new module (and put it into the cache)\n \t\tvar module = installedModules[moduleId] = {\n \t\t\ti: moduleId,\n \t\t\tl: false,\n \t\t\texports: {}\n \t\t};\n\n \t\t// Execute the module function\n \t\tmodules[moduleId].call(module.exports, module, module.exports, __webpack_require__);\n\n \t\t// Flag the module as loaded\n \t\tmodule.l = true;\n\n \t\t// Return the exports of the module\n \t\treturn module.exports;\n \t}\n\n\n \t// expose the modules object (__webpack_modules__)\n \t__webpack_require__.m = modules;\n\n \t// expose the module cache\n \t__webpack_require__.c = installedModules;\n\n \t// define getter function for harmony exports\n \t__webpack_require__.d = function(exports, name, getter) {\n \t\tif(!__webpack_require__.o(exports, name)) {\n \t\t\tObject.defineProperty(exports, name, { enumerable: true, get: getter });\n \t\t}\n \t};\n\n \t// define __esModule on exports\n \t__webpack_require__.r = function(exports) {\n \t\tif(typeof Symbol !== 'undefined' && Symbol.toStringTag) {\n \t\t\tObject.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });\n \t\t}\n \t\tObject.defineProperty(exports, '__esModule', { value: true });\n \t};\n\n \t// create a fake namespace object\n \t// mode & 1: value is a module id, require it\n \t// mode & 2: merge all properties of value into the ns\n \t// mode & 4: return value when already ns object\n \t// mode & 8|1: behave like require\n \t__webpack_require__.t = function(value, mode) {\n \t\tif(mode & 1) value = __webpack_require__(value);\n \t\tif(mode & 8) return value;\n \t\tif((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;\n \t\tvar ns = Object.create(null);\n \t\t__webpack_require__.r(ns);\n \t\tObject.defineProperty(ns, 'default', { enumerable: true, value: value });\n \t\tif(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));\n \t\treturn ns;\n \t};\n\n \t// getDefaultExport function for compatibility with non-harmony modules\n \t__webpack_require__.n = function(module) {\n \t\tvar getter = module && module.__esModule ?\n \t\t\tfunction getDefault() { return module['default']; } :\n \t\t\tfunction getModuleExports() { return module; };\n \t\t__webpack_require__.d(getter, 'a', getter);\n \t\treturn getter;\n \t};\n\n \t// Object.prototype.hasOwnProperty.call\n \t__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };\n\n \t// __webpack_public_path__\n \t__webpack_require__.p = \"/ymzuiku/react-test-renderer-helper.git/\";\n\n\n \t// Load entry module and return exports\n \treturn __webpack_require__(__webpack_require__.s = 2);\n","'use strict';\n\nexports.__esModule = true;\n\nwindow.__testerCache = {};\nwindow.__testerIsTest = true;\n\nvar tester__defaultKey = 'tester__defaultKey';\nfunction tester__setStatus(a, b) {\n  if (typeof window === 'undefined' || !window.__testerIsTest) {\n    return;\n  }\n  var obj, key;\n  if (typeof b === 'undefined') {\n    key = tester__defaultKey;\n    obj = a;\n  } else {\n    key = a;\n    obj = b;\n  }\n  window.__testerCache[key] = obj;\n}\nfunction tester__getStatus(key) {\n  if (typeof key === 'undefined') {\n    key = tester__defaultKey;\n  }\n  return window.__testerCache[key];\n}\nfunction tester__clearStatus(key) {\n  if (typeof key === 'undefined') {\n    key = tester__defaultKey;\n  }\n  if (key === '*') {\n    window.__testerCache = {};\n    tester__default.status = window.__testerCache;\n  } else {\n    delete window.__testerCache[key];\n    tester__default.status = window.__testerCache;\n  }\n}\nvar tester__default = {\n  setStatus: tester__setStatus,\n  getStatus: tester__getStatus,\n  clearStatus: tester__clearStatus,\n  status: window.__testerCache,\n};\n\nexports.default = tester__default;\nmodule.exports = exports['default'];\n","module.exports = require(\"react-test-renderer\");","export default function _defineProperty(obj, key, value) {\n  if (key in obj) {\n    Object.defineProperty(obj, key, {\n      value: value,\n      enumerable: true,\n      configurable: true,\n      writable: true\n    });\n  } else {\n    obj[key] = value;\n  }\n\n  return obj;\n}","import renderer from 'react-test-renderer';\nimport tester from './tester.lib';\nwindow.__testerIsTest = true;\n\nlet rendererType = renderer.create().toJSON();\nrendererType = null;\n\nexport function fixId(ele, id = '') {\n  if (ele && ele.props) {\n    if (id.search(/\\=/) > 0) {\n      const kv = id.split('=');\n      return ele.props[kv[0]] === kv[1];\n    }\n    if (id.search(/\\#/) === 0) {\n      const kv = id.split('#');\n      return ele.props.id === kv[1];\n    }\n    if (id.search(/\\@/) === 0) {\n      const kv = id.split('@');\n      return ele.props.name === kv[1];\n    }\n    if (id.search(/\\./) === 0) {\n      const kv = id.split('.');\n      return ele.props.className === kv[1];\n    }\n    return ele.type === id;\n  }\n  return false;\n}\n\nexport function find(component, id, num = 0) {\n  let child;\n  let jumpNum = -1;\n  function _getChild(comp) {\n    if (comp && comp.children) {\n      for (let i = 0; i < comp.children.length; i++) {\n        if (child) {\n          break;\n        }\n        const ele = comp.children[i];\n        if (fixId(ele, id)) {\n          jumpNum += 1;\n        }\n        if (jumpNum === num) {\n          child = ele;\n        } else {\n          _getChild(ele);\n        }\n      }\n    }\n  }\n  // 如果传入的 component 没有 toJSON，帮它toJSON\n  let rootComp;\n  if (component && component.toJSON) {\n    rootComp = component.toJSON();\n  } else {\n    rootComp = component;\n  }\n  // 首先匹配根节点\n  if (fixId(rootComp, id)) {\n    jumpNum += 1;\n  }\n  if (jumpNum === num) {\n    return rendererType || rootComp;\n  }\n  // 如果根节点没有匹配，递归子节点\n  _getChild(rootComp);\n  return rendererType || child;\n}\n\nexport function getStatus(key) {\n  return tester.getStatus(key);\n}\n\nexport function clearStatus(key) {\n  tester.clearStatus(key);\n}\n\nexport function toSnapshotByHooks(component, id) {\n  if (id === false) {\n    return {\n      props: component.root.props,\n      zip: component.toJSON(),\n    };\n  }\n  return {\n    status: getStatus(id),\n    props: component.root.props,\n    zip: component.toJSON(),\n  };\n}\n\nexport function toSnapshot(component, params) {\n  const instance = component.getInstance();\n  const returnEle = {\n    status: getStatus(component),\n    props: instance.props,\n    state: instance.state,\n    zip: component.toJSON(),\n  };\n  if (params && params.length > 0) {\n    let other = {};\n    for (let i = 0; i < params.length; i++) {\n      other[params[i]] = instance[params[i]];\n    }\n    return {\n      ...returnEle,\n      other,\n    };\n  }\n  return returnEle;\n}\n","import defineProperty from \"./defineProperty\";\nexport default function _objectSpread(target) {\n  for (var i = 1; i < arguments.length; i++) {\n    var source = arguments[i] != null ? arguments[i] : {};\n    var ownKeys = Object.keys(source);\n\n    if (typeof Object.getOwnPropertySymbols === 'function') {\n      ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) {\n        return Object.getOwnPropertyDescriptor(source, sym).enumerable;\n      }));\n    }\n\n    ownKeys.forEach(function (key) {\n      defineProperty(target, key, source[key]);\n    });\n  }\n\n  return target;\n}"],"sourceRoot":""}