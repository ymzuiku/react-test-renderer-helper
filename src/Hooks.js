import React from 'react';
import tester from './lib/tester.lib';

function Hooks() {
  const [num, setNum] = React.useState(0);
  function render() {
    function handleAddNum() {
      setNum(num + 1);
    }
    tester.setStatus({ num });
    return (
      <div>
        <header>
          <h2>{num}</h2>
          <p id="test-click" onClick={handleAddNum}>
            Edit <code>src/App.js</code> and save to reload.
          </p>
          <a href="https://reactjs.org" target="_blank" rel="noopener noreferrer">
            Learn React
          </a>
        </header>
      </div>
    );
  }
  return React.useMemo(render, [num]);
}

export default Hooks;
