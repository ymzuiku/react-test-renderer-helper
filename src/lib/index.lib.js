import renderer from 'react-test-renderer';
import tester from './tester.lib';
window.__testerIsTest = true;

let rendererType = renderer.create().toJSON();
rendererType = null;

export function fixId(ele, id = '') {
  if (ele && ele.props) {
    if (id.search(/\=/) > 0) {
      const kv = id.split('=');
      return ele.props[kv[0]] === kv[1];
    }
    if (id.search(/\#/) === 0) {
      const kv = id.split('#');
      return ele.props.id === kv[1];
    }
    if (id.search(/\@/) === 0) {
      const kv = id.split('@');
      return ele.props.name === kv[1];
    }
    if (id.search(/\./) === 0) {
      const kv = id.split('.');
      return ele.props.className === kv[1];
    }
    return ele.type === id;
  }
  return false;
}

export function find(component, id, num = 0) {
  let child;
  let jumpNum = -1;
  function _getChild(comp) {
    if (comp && comp.children) {
      for (let i = 0; i < comp.children.length; i++) {
        if (child) {
          break;
        }
        const ele = comp.children[i];
        if (fixId(ele, id)) {
          jumpNum += 1;
        }
        if (jumpNum === num) {
          child = ele;
        } else {
          _getChild(ele);
        }
      }
    }
  }
  // 如果传入的 component 没有 toJSON，帮它toJSON
  let rootComp;
  if (component && component.toJSON) {
    rootComp = component.toJSON();
  } else {
    rootComp = component;
  }
  // 首先匹配根节点
  if (fixId(rootComp, id)) {
    jumpNum += 1;
  }
  if (jumpNum === num) {
    return rendererType || rootComp;
  }
  // 如果根节点没有匹配，递归子节点
  _getChild(rootComp);
  return rendererType || child;
}

export function getStatus(key) {
  return tester.getStatus(key);
}

export function clearStatus(key) {
  tester.clearStatus(key);
}

export function toSnapshotByHooks(component, id) {
  if (id === false) {
    return {
      props: component.root.props,
      zip: component.toJSON(),
    };
  }
  return {
    status: getStatus(id),
    props: component.root.props,
    zip: component.toJSON(),
  };
}

export function toSnapshot(component, params) {
  const instance = component.getInstance();
  const returnEle = {
    status: getStatus(component),
    props: instance.props,
    state: instance.state,
    zip: component.toJSON(),
  };
  if (params && params.length > 0) {
    let other = {};
    for (let i = 0; i < params.length; i++) {
      other[params[i]] = instance[params[i]];
    }
    return {
      ...returnEle,
      other,
    };
  }
  return returnEle;
}
