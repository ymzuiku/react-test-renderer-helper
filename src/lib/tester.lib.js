'use strict';

exports.__esModule = true;

window.__testerCache = {};
window.__testerIsTest = true;

var tester__defaultKey = 'tester__defaultKey';
function tester__setStatus(a, b) {
  if (typeof window === 'undefined' || !window.__testerIsTest) {
    return;
  }
  var obj, key;
  if (typeof b === 'undefined') {
    key = tester__defaultKey;
    obj = a;
  } else {
    key = a;
    obj = b;
  }
  window.__testerCache[key] = obj;
}
function tester__getStatus(key) {
  if (typeof key === 'undefined') {
    key = tester__defaultKey;
  }
  return window.__testerCache[key];
}
function tester__clearStatus(key) {
  if (typeof key === 'undefined') {
    key = tester__defaultKey;
  }
  if (key === '*') {
    window.__testerCache = {};
    tester__default.status = window.__testerCache;
  } else {
    delete window.__testerCache[key];
    tester__default.status = window.__testerCache;
  }
}
var tester__default = {
  setStatus: tester__setStatus,
  getStatus: tester__getStatus,
  clearStatus: tester__clearStatus,
  status: window.__testerCache,
};

exports.default = tester__default;
module.exports = exports['default'];
