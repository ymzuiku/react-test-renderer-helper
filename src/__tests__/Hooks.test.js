import 'jest-styled-components';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { toSnapshotByHooks, getStatus, clearStatus, find } from '../lib/index.lib';
import Hooks from '../Hooks';

beforeEach(() => {
  clearStatus('*');
});

test('App 启动不崩溃', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Hooks />, div);
  ReactDOM.unmountComponentAtNode(div);
});

test('App 初始化渲染快照', () => {
  const comp = renderer.create(<Hooks bb={20} />);
  expect(toSnapshotByHooks(comp)).toMatchSnapshot('render');
});

test('App click 10次', () => {
  const comp = renderer.create(<Hooks bb={20} />);
  for (let i = 0; i <  10 ; i++) {
    find(comp, '#test-click').props.onClick();
  }

  expect(toSnapshotByHooks(comp)).toMatchSnapshot();
  expect(getStatus().num).toEqual(10);
});
