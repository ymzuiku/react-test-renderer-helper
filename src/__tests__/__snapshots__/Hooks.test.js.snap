// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App click 10次 1`] = `
Object {
  "props": Object {
    "bb": 20,
  },
  "status": Object {
    "num": 10,
  },
  "zip": <div>
    <header>
      <h2>
        10
      </h2>
      <p
        id="test-click"
        onClick={[Function]}
      >
        Edit 
        <code>
          src/App.js
        </code>
         and save to reload.
      </p>
      <a
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
    </header>
  </div>,
}
`;

exports[`App 初始化渲染快照: render 1`] = `
Object {
  "props": Object {
    "bb": 20,
  },
  "status": Object {
    "num": 0,
  },
  "zip": <div>
    <header>
      <h2>
        0
      </h2>
      <p
        id="test-click"
        onClick={[Function]}
      >
        Edit 
        <code>
          src/App.js
        </code>
         and save to reload.
      </p>
      <a
        href="https://reactjs.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn React
      </a>
    </header>
  </div>,
}
`;
